mpmath/ctx_mp.py

```diff
@@ -1113,21 +1117,25 @@ def nint_distance(ctx, x):
             im_dist = ctx.ninf
         elif hasattr(x, "_mpc_"):
             re, im = x._mpc_
-            isign, iman, iexp, ibc = im
+            try:
+                iman, iexp = to_man_exp(im)
+            except ValueError:
+                raise ValueError("requires a finite number")
             if iman:
-                im_dist = iexp + ibc
-            elif im == fzero:
-                im_dist = ctx.ninf
+                im_dist = iexp + iman.bit_length()
             else:
-                raise ValueError("requires a finite number")
+                im_dist = ctx.ninf
         else:
             x = ctx.convert(x)
             if hasattr(x, "_mpf_") or hasattr(x, "_mpc_"):
                 return ctx.nint_distance(x)
             else:
                 raise TypeError("requires an mpf/mpc")
-        sign, man, exp, bc = re
-        mag = exp+bc
+        try:
+            man, exp = to_man_exp(re)
+        except ValueError:
+            raise ValueError("requires a finite number")
```

Comment:
Considering the number of similar try-except blocks in this PR, would it make sense to introduce a `isfinite` function to mpmath/libmp/libmpf.py:
```python
def isfinite(s):
    sign, man, exp, bc = s
    return (not man) and exp
```
so that here and elsewhere one can use
```python
if not isfinite(re):
    raise ValueError(...)
man, exp = to_man_exp(re)
```

Reply:
In fact, there is such a function (on context instance).  I'll refactor 1st commit, thanks.

Reply:
I hope, the 3rd commit fixes this issue.